File: bindings/java/src/main/java/io/github/everanium/itb3/Status.java
// Status codes mirrored from the libitb3 C ABI
// (cmd/cshared/internal/capi/errors.go). Numeric values are stable
// across releases. The enum is the structural code only: the human
// wording of a failure arrives already composed in the library's own
// diagnostic, so no wording is restated here.

package io.github.everanium.itb3;

/** Integer status code returned by every libitb3 entry point. */
public enum Status {
    OK(0),
    BAD_HASH(1),
    BAD_KEY_BITS(2),
    BAD_HANDLE(3),
    BAD_INPUT(4),
    BUFFER_TOO_SMALL(5),
    ENCRYPT_FAILED(6),
    DECRYPT_FAILED(7),
    SEED_WIDTH_MIX(8),
    BAD_MAC(9),
    MAC_FAILURE(10),
    BLOB_MALFORMED_RECIPE(11),
    RECIPE_PRIMITIVE_UNKNOWN(12),
    UNKNOWN_PROFILE(13),
    RESERVED_14(14),
    RESERVED_15(15),
    RESERVED_16(16),
    RESERVED_17(17),
    BLOB_MODE_MISMATCH(19),
    BLOB_MALFORMED(20),
    BLOB_VERSION_TOO_NEW(21),
    BLOB_TOO_MANY_OPTS(22),
    STREAM_TRUNCATED(23),
    STREAM_AFTER_FINAL(24),
    TRIPLE_CLOSED(25),
    PROFILE_EXISTS(26),
    INTERNAL(99);

    private final int code;

    Status(int code) {
        this.code = code;
    }

    /** The numeric ABI code. */
    public int code() {
        return code;
    }

    /** Maps a raw ABI code onto the enum; an unknown code maps to
     * {@link #INTERNAL} (the raw value stays available on the
     * {@link ItbException} that carries it). */
    public static Status fromCode(int code) {
        for (Status s : values()) {
            if (s.code == code) {
                return s;
            }
        }
        return INTERNAL;
    }
}

File: bindings/clojure/src/io/github/everanium/itb3/clojure/status.clj
(ns io.github.everanium.itb3.clojure.status
  "Status codes mirrored from the libitb3 C ABI
  (cmd/cshared/internal/capi/errors.go), modelled as namespaced-free
  keywords so call sites match with plain `=` / `case`. Numeric
  values are stable across releases; a code outside the known roster
  (a future libitb3 release) maps to :unknown while the raw code stays
  available in the error map.

  The keywords are the structural code only. The human wording of a
  failure arrives already composed in the library's own diagnostic —
  the class of failure and, where there is one, the specific case —
  so no wording is restated here.")

(def ^:private code-table
  [[0  :ok]
   [1  :bad-hash]
   [2  :bad-key-bits]
   [3  :bad-handle]
   [4  :bad-input]
   [5  :buffer-too-small]
   [6  :encrypt-failed]
   [7  :decrypt-failed]
   [8  :seed-width-mix]
   [9  :bad-mac]
   [10 :mac-failure]
   [11 :blob-malformed-recipe]
   [12 :recipe-primitive-unknown]
   [13 :unknown-profile]
   [14 :reserved-14]
   [15 :reserved-15]
   [16 :reserved-16]
   [17 :reserved-17]
   [19 :blob-mode-mismatch]
   [20 :blob-malformed]
   [21 :blob-version-too-new]
   [22 :blob-too-many-opts]
   [23 :stream-truncated]
   [24 :stream-after-final]
   [25 :triple-closed]
   [26 :profile-exists]
   [99 :internal]])

(def ^:private code->kw
  (into {} (map (fn [[code kw]] [code kw])) code-table))

(def ^:private kw->code
  (into {} (map (fn [[code kw]] [kw code])) code-table))

(defn code->status
  "Maps a raw libitb3 status code to its keyword; an unknown code maps
  to :unknown (keep the raw code alongside when reporting)."
  [code]
  (get code->kw code :unknown))

(defn status->code
  "Maps a status keyword back to its numeric ABI code; nil for
  :unknown and unrecognised keywords."
  [status]
  (get kw->code status))
